src/storage/mvcc/txn.rs

```diff
@@ -485,15 +485,20 @@ impl<S: Snapshot> MvccTxn<S> {
     }
 
     pub fn rollback(&mut self, key: Key) -> Result<bool> {
-        self.cleanup(key, 0)
+        // There are only two cases when a key is rolled back:
+        // 1) A failed transaction rolls back its prewritten keys
+        // 2) Other transactions roll back the secondary locks
```

Comment:
Other transactions also can rollback the primary lock.

Reply:
In which case? The primary lock should be only cleaned up when resolving locks.

Reply:
If other transactions try to read a key but found it is locked, this transaction will wait until timeout and try to rollback the transaction which hold the lock.

Reply:
Yes. Rollback the transaction must be finished by **cleanup** the primary lock first.

Reply:
Alright, it strongly depends on the behavior of ti-client. I think you should describe the behavior or constraint of the implementation of client.

Reply:
Sadly, we don't even have a public spec for transactional tikv. :(